### DISCUSSION TITLE
Incrementing a number

### DISCUSSION THREAD

**Discussion by kelvinz** [supabase/supabase#909] [Questions] [answered]
Is there a way to increment a number in a field? Or the only way is to read the field, +1 locally, then update the field?

**Comment by kiwicopple** [comment, marked as answer]
Hey @kelvinz - this should be doable with a stored procedure:

```sql
create function increment () 
returns void as
$$
  update table_name 
  set field_name = field_name + 1; -- increments is done here
$$ 
language sql stable;
```

You can call this from your browser:

```js
const { data, error } = await supabase
  .rpc('increment')
```

Let me know if you get stuck

**Reply by kelvinz** [reply]
Hi @kiwicopple, sorry for the noob-ness, but how do I create the function?

**Reply by steve-chavez** [reply]
@kelvinz You create it on the SQL Editor(click the terminal icon or go to `/editor/sql` ) on the Supabase UI.

**Reply by kelvinz** [reply]
Really sorry, am still having a few issues.
1. How do I update or delete the function, tried editing in Supabase UI, but when I tried running it in the UI, it says that the function already exists so it throws an error.
2. How do I pass in multiple parameters, aka I want to add x rather than always +1.
3. How do I only target the row with id of Y. I don't see any indication of which row the field is being updated in the example.
4. Also when I call the function from my browser it returns a 400 error.

**Reply by silentworks** [reply]
@kelvinz it would have helped if you stated your requirements in your initial question. Here are some answers to some of these.

1. To delete the function use `DROP FUNCTION public.increment;`
2. You can create a function with a parameter and also pass the parameter when calling the function

```sql
create function increment (x int) 
returns void as
$$
  update table_name 
  set field_name = field_name + x; -- increments is done here
$$ 
language sql stable;
```

You call it from the browser like

```js
const { data, error } = await supabase
  .rpc('increment', { x: 1 })
```

3. You can pass filters while using stored procedures, so you should be able to do

```js
const { data, error } = await supabase
  .rpc('increment', { x: 1 })
  .eq('id', 2)
```

4. Not sure why this is not working, but try recreating the function and see if it makes a difference

You can also take a look at the RPC docs to see more about how it works and what you can do with it https://supabase.io/docs/reference/javascript/rpc

Also note, I haven't tested this code but it should work.

**Reply by kelvinz** [reply]
@silentworks Sorry, initially I didn't know about rpcs and didn't know about the possibilities. Am only starting to think of more options as I started playing around with it.

I tried implementing the new function as suggested, but it throws an error
```supabase.rpc(...).eq is not a function```

Wish the rpc docs could include some details on creating the functions for those not familiar with SQL editing. I will blog about what I've learned so far to pass the help forward~

**Reply by silentworks** [reply]
Ah you are right, the docs does say that stored procedures that return a table can use the `.eq`, it didn't say it can be used with an update which returns a void. So maybe you will have to pass the id as a parameter to the rpc function.

```js
const { data, error } = await supabase
  .rpc('increment', { x: 1, id: 2 })
```
and also update your increment stored function to

```sql
create function increment (x int, id int) 
returns void as
$$
  update table_name 
  set field_name = field_name + x; -- increments is done here
  where id = id
$$ 
language sql stable;
```

**Reply by kelvinz** [reply]
Hmm... I tried this new function and it returns a 400 error. 🤷🏻‍♀️

**Reply by cloudorbush** [reply]
stuck on this as well

How difficult can implementing a counter be 😞

**Reply by soedirgo** [reply]
Hey @kelvinz @10000multiplier, sorry this is more difficult than necessary.

The problem was the function was created with `language sql stable`, whereas functions that modify the database need to be `volatile`. Here's the updated function:

```plpgsql
create function increment (x int, id int) 
returns void as
$$
  update table_name 
  set field_name = field_name + x
  where id = id
$$ 
language sql volatile;
```

You might also need to run:

```plpgsql
alter table table_name replica identity full;
```

to allow updates in the first place (we're working on a change so you don't have to do this for future projects).

I've tested this just now on a new project. Hope this helps!

**Reply by kelvinz** [reply]
Hi @soedirgo,

Thank you for your help, can I check what does ```alter table table_name replica identity full;``` do and what is its implications?

**Reply by soedirgo** [reply]
Currently, Supabase's Postgres database has a [publication](https://www.postgresql.org/docs/current/logical-replication.html) named `supabase_realtime` by default (this is necessary for the realtime subscription feature) that publishes insert/update/delete events for all tables.

However, to publish update/delete events, it needs a way to identify the rows. If you have a primary key, you don't need to run the above, e.g. if you use the defaults when creating your table (considering you mentioned about using the `id` column, I'm guessing you are). Otherwise when you do an `UPDATE` Postgres will complain: `cannot update table "table_name" because it does not have a replica identity and publishes updates`

**Reply by cloudorbush** [reply]
> Hey @kelvinz @10000multiplier, sorry this is more difficult than necessary.
> 
> The problem was the function was created with `language sql stable`, whereas functions that modify the database need to be `volatile`. Here's the updated function:
> 
> ```sql
> create function increment (x int, id int) 
> returns void as
> $$
>   update table_name 
>   set field_name = field_name + x
>   where id = id
> $$ 
> language sql volatile;
> ```
> 
> You might also need to run:
> 
> ```sql
> alter table table_name replica identity full;
> ```
> 
> to allow updates in the first place (we're working on a change so you don't have to do this for future projects).
> 
> I've tested this just now on a new project. Hope this helps!

Thanks.

Experiencing a bug where 4 different rows gets updated at the same time, but I'm targeting only one id as in the example. 😕

**Reply by silentworks** [reply]
@10000multiplier this is due to the `id = id`, please change the variable name to something other than id like below.

```sql
create function increment (x int, row_id int) 
returns void as
$$
  update table_name 
  set field_name = field_name + x
  where id = row_id
$$ 
language sql volatile;
```
And call it like

```js
const { data, error } = await supabase
  .rpc('increment', { x: 1, row_id: 2 })
```

I didn't need the alter table code as it just worked without it. Here is a video of me testing this out with it working. 

https://user-images.githubusercontent.com/79497/112918131-b3078800-90f3-11eb-921b-60dba19980fd.mp4

**Reply by cloudorbush** [reply]
> @10000multiplier this is due to the `id = id`, please change the variable name to something other than id like below.
> 
> ```sql
> create function increment (x int, row_id int) 
> returns void as
> $$
>   update table_name 
>   set field_name = field_name + x
>   where id = row_id
> $$ 
> language sql volatile;
> ```
> 
> And call it like
> 
> ```js
> const { data, error } = await supabase
>   .rpc('increment', { x: 1, row_id: 2 })
> ```
> 
> I didn't need the alter table code as it just worked without it. Here is a video of me testing this out with it working.
> 
>  Kapture.2021-03-30.at.01.00.21.mp4

Got it to work. Thank you!

**Reply by kelvinz** [reply]
Thank you all for the answers! Managed to get it to work too!

**Comment by deluxefx** [comment]
@kiwicopple works amazingly

**Comment by wcooperscott** [comment]
If you want to specify what table and field to update as well, here's what I am using:

```
CREATE OR REPLACE FUNCTION increment(table_name TEXT, row_id INT, x INT, field_name TEXT) 
RETURNS VOID AS
$func$
BEGIN
   EXECUTE format('UPDATE %I SET %I = %I + $1 WHERE id = $2', table_name, field_name, field_name)
   USING x, row_id;
END
$func$ LANGUAGE plpgsql;
```

The function takes four parameters:

- `table_name`: the name of the table to update
- `row_id`: the id of the row to update
- `x`: the amount to increment the field value by
- `field_name`: the name of the field to update


Note:
This function should be used as a utility for backend incrementing.  Keep this server-side. Never allow a user to alter what table, field, or value to update.

**Reply by bbernag** [reply]
this is great, thank you so much!

**Reply by Falci** [reply]
feels like reinventing the wheel

**Comment by haffi96** [comment]
Inspired from @wcooperscott 's suggestion. Might be helpful for someone:

Notes:
- I have more than one column which can be chosen to be incremented 
- add a new row in the table if it doesn't exist but if its does, then increment. 

Here is what im using for this: 

```sql
CREATE OR REPLACE FUNCTION increment(user_id_to_check INT, field_name TEXT)
RETURNS VOID AS
$func$
BEGIN
    EXECUTE format('INSERT INTO metric (user_id, %I)
    VALUES ($1, 1) ON CONFLICT (user_id) DO UPDATE
    SET %I = metric.%I + 1 WHERE metric.user_id = $1', field_name, field_name, field_name, field_name)
    USING user_id_to_check;
END
$func$ LANGUAGE plpgsql;
```

I think for this to work, you need a unique constraint on `user_id` (or whichever column you're using to check if the row exists)

To invoke the function:

```typescript
const { data, error } = await supabase.rpc('increment', { user_id_to_check: userID, field_name: field_to_increment })
```

**Comment by timdiacon** [comment]
Can anyone help me work out how to do this but then also return the newly incremented value back to the front end?

**Reply by steve-chavez** [reply]
Should be like:

```sql
create function increment () 
returns bigint as
$$
  update table_name 
  set field_name = field_name + 1; -- increments is done here
  returning field_name -- you can put more fields here, comma separated
$$ 
language sql stable;
```

**Reply by organicnz** [reply]
Here's our use case, but not an expert lol :)

https://dev.to/organicnz/count-and-discount-raws-from-the-likes-table-36jj

**Comment by elschilling** [comment]
I have a table that stores parameters to create dynamic pages and a column to count the number of views of each page (or row in the table). I have RLS policies to restrict insert acess only for authenticated users. Following the steps described here I managed to get the increment function working but had to disable RLS policies for it to work. Is it possible to create an update RLS policy to allow public access only to the 'views' column of the table?

**Reply by wcooperscott** [reply]
I would write a stored procedure named `view(page_id INT)`. This procedure should have a security definer set to override RLS. `view` should increment the page view by 1. 

A better, more honest implementation should limit how many times a person can view a video - preventing view botting. To do this, I would have an extra table `view_info`. View info would have an `id (incremental)`, `ip`, `page_id`, `timestamp`.  Write a stored procedure `view(page_id INT)`. `view` should insert a new row into the `view_info` table - capturing the user's `ip` and the viewed `page_id`. The `view_info` table would have a trigger, called something like `on_view_insert` that adds a view to page (in the `page` table). However, `on_view_insert` should check if more than X rows in `view_info` have the same `ip` and `page_id` as the inserted row. After more than X matches, you may want to prevent the `on_view_insert` trigger from incrementing the page. 

Using the more complex method, you can limit an IP address from giving a page more than 3 views (for example) and keep track of the time when an IP address visited a page. If your users are signed in, it may be better to use their UUID instead of their IP address as an identifier.
